applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,722 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/uart.h>
+#include <misc/byteorder.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#include "ble_controller_hci_vs.h"
+
+#include "chmap_filter.h"
+
+#define MODULE ble_qos
+#include "module_state_event.h"
+#include "ble_event.h"
+#include "config_event.h"
+#include "hid_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_QOS_LOG_LEVEL);
+
+#define INVALID_BLACKLIST 0xFFFF
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+# if DT_NORDIC_NRF_USBD_USBD_0_NUM_IN_ENDPOINTS < 4
+# error Too few USB IN Endpoints enabled. \
+	Modify appropriate dts.overlay to increase num-in-endpoints to 4 or more
+# endif
+#endif
+
+#ifndef ROUNDED_DIV
+#define ROUNDED_DIV(a, b)  (((a) + ((b) / 2)) / (b))
+#endif /* ROUNDED_DIV */
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+#define THREAD_STACK_SIZE		     \
+	(CONFIG_DESKTOP_BLE_QOS_STACK_SIZE + \
+	 CONFIG_DESKTOP_BLE_QOS_STATS_PRINT_STACK_SIZE)
+#else
+#define THREAD_STACK_SIZE CONFIG_DESKTOP_BLE_QOS_STACK_SIZE
+#endif
+
+#define THREAD_PRIORITY K_PRIO_PREEMPT(K_LOWEST_APPLICATION_THREAD_PRIO)
+
+static K_THREAD_STACK_DEFINE(thread_stack, THREAD_STACK_SIZE);
+static struct k_thread thread;
+
+struct params_ble {
+	u16_t sample_count_min;
+	u8_t min_channel_count;
+	s16_t weight_crc_ok;
+	s16_t weight_crc_error;
+	u16_t ble_block_threshold;
+	u8_t eval_max_count;
+	u16_t eval_duration;
+	u16_t eval_keepout_duration;
+	u16_t eval_success_threshold;
+} __packed;
+
+struct params_wifi {
+	s16_t wifi_rating_inc;
+	s16_t wifi_present_threshold;
+	s16_t wifi_active_threshold;
+} __packed;
+
+struct params_chmap {
+	u8_t chmap[5];
+} __packed;
+
+struct params_blacklist {
+	u16_t wifi_chn_bitmask;
+} __packed;
+
+static struct chmap_instance *chmap_inst;
+static u8_t current_chmap[5] = { 0xFF, 0xFF, 0xFF, 0xFF, 0x1F };
+static atomic_t processing;
+static atomic_t new_blacklist;
+static atomic_t params_updated;
+static struct chmap_filter_params filter_params;
+static struct k_mutex data_access_mutex;
+
+BUILD_ASSERT(sizeof(struct bt_hci_cp_le_set_host_chan_classif) ==
```

Comment:
This is possible, but there is still disconnect between this code, python script, and .lib. If bt_hci_cp_le_set_host_chan_classif were to be refactored, this would become a silent mismatch with python script and .lib instead of build error.

Reply:
Ok leave it then if the library has to be decoupled from BLE stack.